src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+namespace drawing_recognition {
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+#define MAX(a,b) (((a)>(b))?(a):(b))
+
+struct Point {
+    float x;
+    float y;
+};
+
+struct Line {
+    // ax + by + c = 0
+    float slope;
+    float a;
+    float b;
+    float c;
+};
+
+void make_point(Point *P, const flex_dict &point_dict) {
+    flex_string x ("x");
+    for (auto key_value_pair: point_dict) {
+        const flex_string &key = key_value_pair.first.get<flex_string>();
+        const flex_float &value = key_value_pair.second.get<flex_float>();
+        if (key.compare(x) == 0) {
+            P->x = value;
+        } else {
+            P->y = value;
+        }
+    }
+}
+
+float line_to_point_distance(Line *L, const flex_dict &point) {
+    Point P; make_point(&P, point);
+    int x = (int)(P.x);
+    int y = (int)(P.y);
+    float a = L->a; float b = L->b; float c = L->c;
+    float distance = (float)(fabs(a*x+b*y+c))/sqrt(a*a+b*b);
+    return (int)distance;
+}
+
+void initialize_line(Line *L, const flex_dict &p1, const flex_dict &p2) {
+    Point P1; make_point(&P1, p1);
+    Point P2; make_point(&P2, p2);
+    float start_x = P1.x;
+    float start_y = P1.y;
+    float end_x = P2.x;
+    float end_y = P2.y;
+    if (start_x == start_y) {
+        L->slope = INT_MAX;
+    } else {
+        L->slope = ((float)(end_y - start_y))/((float)(end_x - start_x));
+    }
+    L->a = L->slope;
+    L->b = -1;
+    L->c = start_y - L->slope * start_x;
+}
+
+template<typename T>
+std::vector<T> slice(std::vector<T> const &v, int m, int n) {
+    auto first = v.cbegin() + m;
+    auto last = v.cbegin() + n + 1;
+    std::vector<T> vec(first, last);
+    return vec;
+}
+
+flex_list ramer_douglas_peucker(flex_list stroke, float epsilon) {
+    float dmax = 0;
+    int index = 0;
+    flex_list compressed_stroke;
+    int num_points = stroke.size();
+    Line L;
+    const flex_dict &first_point = stroke[0].get<flex_dict>();
+    const flex_dict &last_point = stroke[num_points-1].get<flex_dict>();
+    initialize_line(&L, first_point, last_point);
+    for (int i = 0; i < num_points; i++) {
+        float d = line_to_point_distance(&L, stroke[i].get<flex_dict>());
+        if (d > dmax) {
+            index = i;
+            dmax = d;
+        }
+    }
+    if (dmax > epsilon) {
+        flex_list first_slice = slice(stroke, 0, index-1);
+        flex_list last_slice = slice(stroke, index, num_points-1);
+        compressed_stroke = ramer_douglas_peucker(first_slice, epsilon);
+        flex_list rec_results_2 = ramer_douglas_peucker(last_slice, epsilon);
+        compressed_stroke.insert(compressed_stroke.end(), 
+            rec_results_2.begin(), rec_results_2.end());
+    } else {
+        compressed_stroke.push_back(first_point);
+        compressed_stroke.push_back(last_point);
+    }
+    return compressed_stroke;
+}
+
+flex_list simplify_drawing(flex_list raw_drawing) {
+    size_t num_strokes = raw_drawing.size();
+    size_t min_x = INT_MAX;
```

Comment:
In C++ code, prefer [std::numeric_limits<size_t>](https://en.cppreference.com/w/cpp/types/numeric_limits).

In particular, note that `INT_MAX` is not the maximum value that `size_t` can store